Out-of-line construction of OperatorName.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32168 Eliminate exception throwing code from dispatch call sites
* **#32121 Out-of-line construction of OperatorName.**
* #32118 Move error reporting code out-of-line from header.
* #32117 Make an assert on a hotpath trigger only in DEBUG mode.

This reduces code size in the call sites of this function (of which
there are many: one for every operator call) since we no longer have
to construct std::string at the site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392951](https://our.internmc.facebook.com/intern/diff/D19392951)

ezyang committed:
Out-of-line construction of OperatorName.

This reduces code size in the call sites of this function (of which
there are many: one for every operator call) since we no longer have
to construct std::string at the site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ea8ac870:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ea8ac870ec9ad8f4d721d17c606588a43d729e3d).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4223190) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84386742) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84386742))
<details>
<summary>
<code>Jan 13 20:56:14 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 13 20:56:14 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 13 20:56:14 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 20:56:14 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 20:56:14 1 out of 1 hunk FAILED 
Jan 13 20:56:14 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 20:56:14 Traceback (most recent call last): 
Jan 13 20:56:14   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 13 20:56:14     patch_repo(args) 
Jan 13 20:56:14   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 13 20:56:14     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 13 20:56:14 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 20:56:14 + cleanup 
Jan 13 20:56:14 + retcode=1 
Jan 13 20:56:14 + set +x 
Jan 13 20:56:14 =================== sccache compilation log =================== 
Jan 13 20:56:14 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 13 20:56:14 Compile requests                 1 
Jan 13 20:56:14 Compile requests executed        0 
Jan 13 20:56:14 Cache hits                       0 
Jan 13 20:56:14 Cache misses                     0 
Jan 13 20:56:14 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer reviewed (approved):
Looks good. Have you measured the impact?

ezyang commented:
> Looks good. Have you measured the impact?

Only measured by eyeballing the generated assembly.

ezyang committed:
Update on "Out-of-line construction of OperatorName."


This reduces code size in the call sites of this function (of which
there are many: one for every operator call) since we no longer have
to construct std::string at the site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392951](https://our.internmc.facebook.com/intern/diff/D19392951)

[ghstack-poisoned]

ezyang committed:
Update on "Out-of-line construction of OperatorName."


This reduces code size in the call sites of this function (of which
there are many: one for every operator call) since we no longer have
to construct std::string at the site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392951](https://our.internmc.facebook.com/intern/diff/D19392951)

[ghstack-poisoned]
